src/Microsoft.TemplateEngine.Cli/HelpAndUsage/HelpForTemplateResolution.cs

```diff
@@ -302,55 +293,40 @@ private static void DisplayParametersInvalidForSomeTemplates(IReadOnlyList<strin
         }
 
-        private static void DisplayPartialNameMatchLanguageAndContextProblems(string templateName, string templateLanguage, string context, TemplateListResolutionResult templateResolutionResult, out bool shouldShowTemplateList)
+        private static void DisplayPartialNameMatchLanguageAndContextProblems(string templateName, string templateLanguage, string context, ListOrHelpTemplateListResolutionResult templateResolutionResult, string baselineName)
         {
-            shouldShowTemplateList = false;
-
-            if (templateResolutionResult.IsNoTemplatesMatchedState || templateResolutionResult.UsingPartialMatches)
+            bool anythingReported = false;
+            if (templateResolutionResult.HasExactMatches)
             {
-                ShowNoTemplatesFoundMessage(templateName, templateLanguage, context);
-                Reporter.Error.WriteLine();
                 return;
             }
-
-            bool anythingReported = false;
-            int partialTemplatesMatchCount = templateResolutionResult.ContextProblemMatchGroups.Count(templateGroup =>
-                {
-                    // all templates in a group should have the same context & name
-                    if (templateGroup[0].Info.Tags != null && templateGroup[0].Info.Tags.TryGetValue("type", out ICacheTag typeTag))
-                    {
-                        MatchInfo? matchInfo = WellKnownSearchFilters.ContextFilter(context)(templateGroup[0].Info);
-                        return ((matchInfo?.Kind ?? MatchKind.Mismatch) == MatchKind.Mismatch);
-                    }
-
-                    // this really shouldn't ever happen. But better to have a generic error than quietly ignore the partial match.
-                    // Cannot retrieve the type for {0}.
-                    Reporter.Error.WriteLine(string.Format(LocalizableStrings.GenericPlaceholderTemplateContextError, templateGroup[0].Info.Name).Bold().Red());
-                    anythingReported = true;
-                    return false;
-                });
-
-            if (partialTemplatesMatchCount > 0)
+            else
             {
-                ShowNoTemplatesFoundMessage(templateName, templateLanguage, context);
-                // {0} template(s) partially matched, but failed on {1}.
-                Reporter.Error.WriteLine(string.Format(LocalizableStrings.TemplatesNotValidGivenTheSpecifiedFilter, partialTemplatesMatchCount, string.Concat("type=", context)).Bold().Red());
+                //
+                ShowNoTemplatesFoundMessage(templateName, templateLanguage, context, baselineName);
                 anythingReported = true;
             }
 
-            if (templateResolutionResult.RemainingPartialMatchGroups.Count > 0)
+            if (templateResolutionResult.HasPartialMatches)
             {
-                shouldShowTemplateList = true;
+                // {0} template(s) partially matched, but failed on {1}.
+                Reporter.Error.WriteLine(
```

Comment:
minor: be aware that string.format is a poor performer when compared to string.concat.  The choice here is maintenance vs. perf.  In cases of simple string interpolation, which under the hood uses string.format, the expression is JITTED. This may offer a better equilibrium between both criterion.

Reply:
this pattern is commonly used in HelpForTemplateResolution and TemplateDetailsDisplay at the moment. It makes sense to refactor it everywhere, but I would suggest separate task for it. What do you think?

Reply:
sounds good.